funsor/distributions.py

```diff
@@ -449,23 +450,79 @@ def eager_mvn(loc, scale_tril, value):
+    assert len(loc.shape) == 1
+    assert len(scale_tril.shape) == 2
+    assert value.output == loc.output
+    if not is_affine(loc) or not is_affine(value):
+        return None  # lazy
+
+    # Extract an affine representation.
+    eye = torch.eye(scale_tril.data.size(-1)).expand(scale_tril.data.shape)
+    prec_sqrt = Tensor(eye.triangular_solve(scale_tril.data, upper=False).solution,
+                       scale_tril.inputs)
+    affine = prec_sqrt @ (loc - value)
+    const, coeffs = extract_affine(affine)
+    if not isinstance(const, Tensor):
+        return None  # lazy
+    if not all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values()):
+        return None  # lazy
+    scale_diag = Tensor(scale_tril.data.diagonal(dim1=-1, dim2=-2),
+                        scale_tril.inputs)
+    log_prob = (-0.5 * scale_diag.shape[-1] * math.log(2 * math.pi)
+                - scale_diag.log().sum() - 0.5 * (const ** 2).sum())
+    print(f"log_prob = {log_prob}")
+
+    # Dovetail to avoid variable name collision in einsum.
+    equations1 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a')) for c in eqn)
+                  for _, eqn in coeffs.values()]
+    equations2 = [''.join(c if c in ',->' else chr(ord(c) * 2 - ord('a') + 1) for c in eqn)
+                  for _, eqn in coeffs.values()]
 
-    dim, = loc.output.shape
-    inputs, (loc, scale_tril) = align_tensors(loc, scale_tril)
-    inputs.update(value.inputs)
-    int_inputs = OrderedDict((k, v) for k, v in inputs.items() if v.dtype != 'real')
+    real_inputs = OrderedDict((k, v) for k, v in affine.inputs.items() if v.dtype == 'real')
+    assert tuple(real_inputs) == tuple(coeffs)
 
-    precision = cholesky_inverse(scale_tril)
-    info_vec = precision.matmul(loc.unsqueeze(-1)).squeeze(-1)
-    log_prob = (-0.5 * dim * math.log(2 * math.pi)
-                - scale_tril.diagonal(dim1=-1, dim2=-2).log().sum(-1)
-                - 0.5 * (loc * info_vec).sum(-1))
-    return Tensor(log_prob, int_inputs) + Gaussian(info_vec, precision, inputs)
+    # Align and broadcast tensors.
+    tensors = [coeff for coeff, _ in coeffs.values()] + [const, scale_diag]
+    inputs, tensors = align_tensors(*tensors, expand=True)
+    coeffs, const, scale_diag = tensors[:-2], tensors[-2], tensors[-1]
+    dim = sum(d.num_elements for d in real_inputs.values())
+    batch_shape = const.shape[:-1]
+
+    info_vec = BlockVector(batch_shape + (dim,))
+    precision = BlockMatrix(batch_shape + (dim, dim))
+    offset1 = 0
+    for i1, (v1, c1) in enumerate(zip(real_inputs, coeffs)):
+        size1 = real_inputs[v1].num_elements
+        slice1 = slice(offset1, offset1 + size1)
+        inputs1, output1 = equations1[i1].split('->')
+        input11, input12 = inputs1.split(',')
+        assert input11 == input12 + output1
+        info_vec[..., slice1] = torch.einsum(
+            f'...{input11},...{output1}->...{input12}', c1, const) \
+            .reshape(batch_shape + (size1,))
```

Comment:
@fritzo I guess we need negative sign here. Everything else looks good to me.

Reply:
Nevermind, you catch it. :D

Reply:
done :smile: